### DISCUSSION TITLE
Annotating TiddlyWiki with Hypothesis

### DISCUSSION THREAD

**Discussion by pesho-ivanov** [TiddlyWiki/TiddlyWiki5#6985] [Ideas]
Hello TiddlyWiki and @Hypothesis communities!

Annotating the web is grand goal (and I am wondering how it will work instead of blog comments), so I want to annotate [my **T**iddly**W**iki](https://pesho-ivanov.github.io/#Binary%20search) using [**H**ypothesis](https://github.com/hypothesis). A year ago I asked about it on the H mailing list ([link](https://groups.google.com/a/list.hypothes.is/g/dev/c/qTKwveU74Ts)) and the conclusion seemed to be that it is a reasonable idea but the client-side navigation of TW makes it problematic to associate the displayed content with the URL and the `<head>`. I would like to involve the TW community in this discussion too since it is not clear to me what the good solution would look like.

Currently H _somewhat_ works for TW but H may display annotations from unopened tiddlers or hide annotations that should be displayed. Since TW supports updating the URL with a section (`url#section`) after an interaction with a tiddler, could accounting for the URL section by H be a solution of the problem or is the <head> element is also required? If so, could TW become more annotation-friendly?

Best,
Pesho

**Comment by pmario** [comment]
You should have linked to the mailing list thread. ... From your post it isn't clear, what meta-data they expect. You mention `<head>`, but that's not specific enough to be useful. 

I saw in your wiki, you did already enable the "Include the target tiddler and the current story sequence" URL setting, which should allow H to open any selection of tiddlers in a defined order of the story-river. ... 

Is this not enough?

**Reply by pesho-ivanov** [reply]
Thanks for the suggestions, @pmario!

1. I link directly to the thread now.
2. I am not sure what exactly H requires -- thus inviting them (incl. @robertknight) to the discussion
3. I now returned to only "target tiddler"

**Comment by pmario** [comment]
@pesho-ivanov ... I did read a bit of the h. docs about [fussy anchoring](https://web.hypothes.is/blog/fuzzy-anchoring/) mentioned in the GG thread. 

That's very similar, to what we can do with the dynannotate plugin in TW itself. .. TW has the diff-match-patch code built into the TW core because we use it to show differences between modified tiddlers and their original tiddler. We also use the code to show differences between existing and imported tiddlers. 

The main problem is, that the UI is completely different to h. .. A bit more investigation would be needed

**Comment by robertknight** [comment]
The root issue here I think is a conceptual mismatch between how Hypothesis expects a "typical" web page or application to work, and the TiddlyWiki model.

Hypothesis' client software has a concept of "integrations" for different document viewers and document types. When you activate Hypothesis on a web page, you get the "HTML" integration by default. The HTML integration assumes a "typical" server or client-rendered web application in which:
- The displayed content is reflected in the path and query string of the URL (the fragment/hash is always ignored)
- There is one logical document per frame, whose content is fully present in the DOM.

This fits most content on the web (blog posts, Wikipedia pages, news articles etc.), but it doesn't look like it fits TiddlyWiki, at least not its default behavior. There are two separate issues:

1. There can be multiple "documents" in the same frame at the same time.
2. Knowing when documents appear or go away requires a specific understanding of TiddlyWiki, since the path and query string of the URL do not update

The assumption that Hypothesis makes of there being one document at a time is fairly deeply embedded in the software, and would require significant work to change. So to make TW work with Hypothesis, you can either:

- Find ways to change TiddlyWiki to make it conform to the model / behaviors that Hypothesis expects
- Ignore the Hypothesis browser software entirely and just use its API for storage (see https://h.readthedocs.io/en/latest/api/)

**Reply by pesho-ivanov** [reply]
Thank you for the extensive explanation, Robert! It seems to me that in the the only missing feature of TW needed to be H-friendly is to handle tiddler URLs without hashes.

Does someone know if there were already approaches/plugins aiming at modifying the URL format to removing the hashes or providing the tiddle-titles as query strings?
